spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+    }
+
+
+
+    fun lastIndexOf (o: Object): int
+    {
+        result = action LIST_FIND(storage, o, 0, length, -1);
+    }
+
+
+    fun clone (): Object
+    {
+        storageCopy = action LIST_DUP(storage);
+        result = new ArrayList(
+            state=self.state, storage=storageCopy, length=self.length);
+    }
+
+
+    fun toArray (): array<Object>
+    {
+        // #problem
+        //How to create a new array correctly ?
+        var a: array<int>;
+        result = action LIST_TO_ARRAY(storage, a, 0, length);
+    }
+
+
+    fun toArray (a: list<Object>): array<Object>
+    {
+        result = action LIST_TO_ARRAY(storage, a, 0, length);
+    }
+
+
+    fun get (index: int): Object
+    {
+        _checkValidIndex(index, length);
+        result = action LIST_GET(storage, index);
+    }
+
+
+    fun set (index: int, element: Object): Object
+    {
+        result = _setElement(index, element);
+    }
+
+
+    fun add (e: Object): boolean
+    {
+        modCount = modCount + 1;
+        action LIST_INSERT_AT(storage, length, e);
+        length = length + 1;
+        result = true;
+    }
+
+
+    fun add (index: int, element: Object): void
+    {
+        _addElement(index, e);
+    }
+
+
+    fun remove (index: int): Object
+    {
+        result = _deleteElement(index);
+    }
+
+
+    fun equals (other: Object): boolean
+    {
+        if (other == self)
+        {
+            result = true;
+        }
+        else
+        {
+            // #problem
+            //We don't know at this moment how create this method.
+            action NOT_IMPLEMENTED();
+        }
+    }
+
+
+    fun hashCode (): int
+    {
+        // result = action OBJECT_HASH_CODE(self);
+        // #problem
+        action NOT_IMPLEMENTED();
+    }
+
+
+    fun remove (o: Object): boolean
+    {
+        var index = action LIST_FIND(storage, o, 0, length, +1);
+        if (index == -1)
+        {
+            result = false;
+        }
+        else
+        {
+            action LIST_REMOVE(storage, index, 1);
+            result = true;
+        }
+    }
+
+
+    fun clear (): void
+    {
+        action LIST_RESIZE(storage, 0);
+        length = 0;
+        modCount = modCount + 1;
+    }
+
+
+    fun addAll (c: Collection): boolean
+    {
+        result = _addAllElements(length, c);
+    }
+
+
+
+    fun addAll (index: int, c: Collection): boolean
+    {
+        _rangeCheckForAdd(index, length);
+        result = _addAllElements(index, c);
+    }
+
+
+    fun removeAll (c: Collection): boolean
+    {
+        // #problem.
+        //I want to create such action:
+
+        //define action SUBTRACTING_SETS(
+        //        aListSubtracting: array<any>,
+        //        aListDeductible: array<any>
+        //    ): list<any>;
+
+        //But Collection isn't List.
+        //But it has method: toArray();
+        //Maybe we must have two arrays as params ?
+
+        //storage = action SUBTRACTING_SETS(storage, c);
```

Comment:
A foreign `Collection` is not a `list` nor an `array`. It's an *interface*.

Reply:
Comment was deleted;